zerver/lib/narrow.py

```diff
@@ -861,6 +873,65 @@ def get_channel_from_narrow_access_unchecked(
     return None
 
 
+# This function corrects the conversation filters of the narrow - channel,
+# topic, DM conversation to point to the narrow in which the message
+# id operand of the `with` operator is present in.
+#
+# This is done as an effort to facilitate permalinks for topics -- if there
+# exists a message represented by the id (`with` operand), we redirect to
+# the narrow in which the message is present. Else, we fallback
+# to the narrow mentioned originally.
+def update_narrow_terms_containing_with_operator(
+    realm: Realm,
+    maybe_user_profile: Union[UserProfile, AnonymousUser],
+    narrow: Optional[List[NarrowParameter]],
+) -> Optional[List[NarrowParameter]]:
+    if narrow is None:
+        return narrow
+
+    contains_with_operator = list(filter(lambda term: term.operator == "with", narrow))
+
+    if len(contains_with_operator) > 0:
+        with_term = contains_with_operator[0]
+        with_term_index = narrow.index(with_term)
+        message_id = int(with_term.operand)
+
+        narrow.remove(with_term)
+
+        if maybe_user_profile.is_authenticated:
+            try:
+                message = access_message(maybe_user_profile, message_id)
+            except JsonableError:
+                return narrow
+        else:
+            try:
+                message = access_web_public_message(realm, message_id)
+            except MissingAuthenticationError:
+                return narrow
+
+        narrow = narrow[with_term_index:]
```

Comment:
This makes an incorrect assumption that the `with` term must have after `channel`/`topic` terms. There are AFAIK no requirements as to ordering of terms in a narrow object; while we have a typical ordering we use for readability, no code should rely on the ordering; instead this should be finding any `topic` operator and any `channel` operator and specifically excising those.

Reply:
updated!